spec/unit/http/session_spec.rb

```diff
@@ -83,4 +83,60 @@ def check_connection?(session, service, ssl_context: nil)
       }.to raise_error(ArgumentError, "Unknown service westbound")
     end
   end
+
+  context 'when resolving using multiple resolvers' do
+    let(:session) { client.create_session }
+
+    it "resolves once per session"
+
+    it "prefers SRV records" do
+      Puppet[:use_srv_records] = true
+      Puppet[:srv_domain] = "srv.example.com"
+
+      expect(session.resolvers.map(&:class)).to eq([Puppet::HTTP::Resolver::SRV, Puppet::HTTP::Resolver::Settings])
+      expect(session.resolvers[0]).to receive(:resolve).and_return(nil).ordered
+      expect(session.resolvers[1]).to receive(:resolve).and_call_original.ordered
+
+      session.route_to(:ca)
+    end
+
+    it "next prefers server_list" do
+      Puppet[:use_srv_records] = true
+      Puppet[:srv_domain] = "srv.example.com"
+      Puppet[:server_list] = 'foo.example.com,bar.example.com'
+
+      expect(session.resolvers.map(&:class)).to eq([Puppet::HTTP::Resolver::SRV, Puppet::HTTP::Resolver::ServerList, Puppet::HTTP::Resolver::Settings])
+      expect(session.resolvers[0]).to receive(:resolve).and_return(nil).ordered
+      expect(session.resolvers[1]).to receive(:resolve).and_return(nil).ordered
+      expect(session.resolvers[2]).to receive(:resolve).and_call_original.ordered
+
+      session.route_to(:ca)
+    end
+
+    it "next prefers settings" do
+      Puppet[:use_srv_records] = true
+      Puppet[:srv_domain] = "srv.example.com"
+      Puppet[:server_list] = 'foo.example.com,bar.example.com'
+      Puppet[:ca_server] = 'server.example.com'
+
+      expect(session.resolvers.map(&:class)).to eq([Puppet::HTTP::Resolver::SRV, Puppet::HTTP::Resolver::ServerList, Puppet::HTTP::Resolver::Settings])
+      expect(session.resolvers[0]).to receive(:resolve).and_return(nil).ordered
+      expect(session.resolvers[1]).to receive(:resolve).and_return(nil).ordered
+      expect(session.resolvers[2]).to receive(:resolve).and_return(URI('server.example.com')).ordered
```

Comment:
May need to return an instance of `Puppet::HTTP::Service::Ca` instead

Reply:
Yeah, I'm stumbling on these tests. It's not super clear to me how these tests are working... I've been reading up on `and_call_original`, but I'm not quite understanding why we want to use that?

Reply:
The `and_call_original` allows the resolver's method to be called, so a proper `Service` object can be returned. If that is omitted, then you'll have to add something like `.and_return(Puppet::HTTP::Service::Ca.new)`. But it's better to not stub out code we control. In this test, we want resolvers 0 and 1 to return nil, and have resolver 2 call the original method. I think the expectation below (resolver 3) can be dropped